File: BEmu/IntradayBarRequest/ElementIntradayBarDateTime.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="IntradayBarRequest/ElementIntradayBarDateTime.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.IntradayBarRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ElementIntradayBarDateTime : Element
    {
        private readonly Datetime _date;

        internal ElementIntradayBarDateTime(DateTime date)
        {
            this._date = new Datetime(date, Datetime.DateTimeTypeEnum.both);
        }

        public override Name  Name { get { return new Name("time"); } }
        public override int NumValues { get { return 1; } }
        public override int NumElements { get { return 0; } }
        public override bool IsComplexType { get { return false; } }
        public override bool IsArray { get { return false; } }
        public override bool IsNull { get { return false; } }
        public override IEnumerable<Element> Elements { get { if (this.NumValues < -1) yield return null; } } //This is the only way I can think to have the compiler output "Enumeration yielded no results"
        public override bool HasElement(string name, bool excludeNullElements = false) { return false; }

        public override object this[int index]
        {
            get
            {
                if (index == 0)
                    return this._date;
                else
                    return base[index];
            }
        }

        public override object GetValue()
        {
            return this._date;
        }

        public override object GetValue(int index)
        {
            if (index == 0)
                return this.GetValue();
            else 
                return base.GetValue(index);
        }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            return base.PrettyPrintHelper(tabIndent, this._date.ToSystemDateTime().ToString("yyyy-MM-ddTHH:mm:ss.fff"));
        }

        public override Datetime GetValueAsDate()
        {
            return this._date;
        }

        public override Datetime GetValueAsDate(int index)
        {
            if (index == 0)
                return this.GetValueAsDate();
            else
                return base.GetValueAsDate(index);
        }

        public override Datetime GetValueAsDatetime(int index)
        {
            if (index == 0)
                return this.GetValueAsDatetime();
            else
                return base.GetValueAsDatetime(index);
        }

        public override Datetime GetValueAsDatetime()
        {
            return this._date;
        }

        public override Datetime GetValueAsTime()
        {
            return this._date;
        }

        public override Datetime GetValueAsTime(int i)
        {
            if (i == 0)
                return this.GetValueAsTime();
            else
                return base.GetValueAsTime(i);
        }

        public override string GetValueAsString()
        {
            return this._date.ToString();
        }

        public override string GetValueAsString(int i)
        {
            if (i == 0)
                return this.GetValueAsString();
            else
                return base.GetValueAsString(i);
        }
    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/IntradayBarRequest/ElementIntradayBarDateTime.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.IntradayBarRequest.ElementIntradayBarDateTime.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu.IntradayBarRequest;

import com.bemu.BEmu.Element;
import com.bemu.BEmu.Name;
import com.bemu.BEmu.Datetime;
import java.text.SimpleDateFormat;
import java.lang.StringBuilder;

public class ElementIntradayBarDateTime extends Element
{
    private final Datetime _date;
    
    ElementIntradayBarDateTime(Datetime date)
    {
    	this._date = date;
    }
    
    public Name name()
    {
    	return new Name("time");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 0;
    }
    
    public boolean isComplexType()
    {
    	return false;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    public boolean hasElement(String name)
    {
    	return false;
    }
    
    public boolean hasElement(String name, boolean excludeNullElements)
    {
    	return false;
    }
    
    protected StringBuilder prettyPrint(int tabIndent)
    {
    	return super.prettyPrintHelper(tabIndent, this.getValueAsString());
    }
    
    public Object getValue()
    {
    	return this._date;
    }
    
    public Object getValue(int index) throws Exception
    {
    	if(index == 0)
    		return this.getValue();
    	else
    		return super.getValue();
    }
    
    public Datetime getValueAsDate()
    {
        return this._date;
    }

    public Datetime getValueAsDate(int index) throws Exception
    {
        if (index == 0)
            return this.getValueAsDate();
        else
            return super.getValueAsDate(index);
    }

    public Datetime getValueAsDatetime(int index) throws Exception
    {
        if (index == 0)
            return this.getValueAsDatetime();
        else
            return super.getValueAsDatetime(index);
    }

    public Datetime getValueAsDatetime()
    {
        return this._date;
    }

    public Datetime getValueAsTime()
    {
        return this._date;
    }

    public Datetime getValueAsTime(int i) throws Exception
    {
        if (i == 0)
            return this.getValueAsTime();
        else
            return super.getValueAsTime(i);
    }

    public String getValueAsString()
    {
    	SimpleDateFormat df = new SimpleDateFormat();
    	df.applyPattern("yyyy-MM-dd'T'HH:mm:ss.SSS");
    	String strDate = df.format(this._date.calendar().getTime());
    	return strDate;
    }

    public String getValueAsString(int i) throws Exception
    {
        if (i == 0)
            return this.getValueAsString();
        else
            return super.getValueAsString(i);
    }
}
